datafusion/functions/src/core/getfield.rs

```diff
@@ -107,29 +109,55 @@ impl ScalarUDFImpl for GetFieldFunc {
                 );
             }
         };
+
         match (array.data_type(), name) {
-                (DataType::Map(_, _), ScalarValue::Utf8(Some(k))) => {
-                    let map_array = as_map_array(array.as_ref())?;
-                    let key_scalar = Scalar::new(StringArray::from(vec![k.clone()]));
-                    let keys = arrow::compute::kernels::cmp::eq(&key_scalar, map_array.keys())?;
-                    let entries = arrow::compute::filter(map_array.entries(), &keys)?;
```

Comment:
using filter will reduce the number of input rows to the number of rows that have keys matching the input key. But we want to respect the number of input rows, and give null for any rows not having the matching key

Reply:
I don't understand this

If the input is like this (two rows, each three elements)
```
{ a: 1, b: 2, c: 100}
{ a: 3, b: 4, c: 200}
```

An expression like `col['c']` will still return 2 rows (but each row will have only a single element)

```
{ c: 100 }
{ c: 200 }
```

Reply:
**Previous implememtation**

map_array.entries() has type of
```
pub struct StructArray {
    len: usize,
    data_type: DataType,
    nulls: Option<NullBuffer>,
    fields: Vec<ArrayRef>,
}
```
With the example above, the layout of "field" will be a vector of 2 array, where first array is a list of key, and second array is a list of value
```
[0]: ["a","b","c","a","b",c"]
[1]: [1,2,100,3,4,200]
```
```
                    let keys = arrow::compute::kernels::cmp::eq(&key_scalar, map_array.keys())?;
```
with this computation, the result is a boolean aray where "key" = "c"
```
[false,false,true,false,false,true]
```
 and thus this operation will reduce the number of rows into
```
                    let entries = arrow::compute::filter(map_array.entries(), &keys)?;
```
```
[0]: ["c,"c"]
[1]: [100,200]
```

**Problem**

However, let's add a row where the map does not have key "c" in between
```
{ a: 1, b: 2, c: 100}
{ a: 1, b: 2}
{ a: 3, b: 4, c: 200}
```
map_array.entries() underneath is represented as
```
[0]: ["a,"b","c","a","b","a","b","c"]
[1]: [1,2,100,1,2,3,4,200]

                    let entries = arrow::compute::filter(map_array.entries(), &keys)?;
Now rows after filtered will be
[0]: ["c","c"]
[1]: [100,200]
```
and the return result will be 
```
{ c: 100 }
{ c: 200 }
```
instead of
```
{ c: 100 }
null
{ c: 200 }
```

Reply:
I would expect the result of evaluating `col[b]` on

```
{ a: 1, b: 2, c: 100}
{ a: 1, b: 2}
{ a: 3, b: 4, c: 200}
```

to be:
```
{ c: 100 }
null
{ c: 200 }
```

For example, in duckdb:

```sql
D create table foo as values (MAP {'a':1, 'b':2, 'c':100}), (MAP{ 'a':1, 'b':2}), (MAP {'a':1, 'b':2, 'c':200});
D select * from foo;
┌───────────────────────┐
│         col0          │
│ map(varchar, integer) │
├───────────────────────┤
│ {a=1, b=2, c=100}     │
│ {a=1, b=2}            │
│ {a=1, b=2, c=200}     │
└───────────────────────┘
D select col0['c'] from foo;
┌───────────┐
│ col0['c'] │
│  int32[]  │
├───────────┤
│ [100]     │
│ []        │
│ [200]     │
└───────────┘
```

Basically a scalar function has the invarant that each input row produces exactly 1 output row

Reply:
yes, the previous implementation will not return null (according to slt https://github.com/apache/datafusion/pull/10148/files#diff-4db02ec06ada20062cbb518eeb713c2643f5a51e89774bdadd9966410baa7d1dR47)

Reply:
i also explained in this discussion: https://github.com/apache/datafusion/pull/10148#discussion_r1580329368